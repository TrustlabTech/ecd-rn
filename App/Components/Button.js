/**
 * Early Childhood Development App
 * @copyright 2016 Global Consent Ltd
 * Civvals, 50 Seymour Street, London, England, W1H 7JG
 * @author Werner Roets <werner@io.co.za>
 */

import React, { Component } from 'react'
import {
  View,
  Text,
  TouchableHighlight,
  StyleSheet,
  ToastAndroid,
  InteractionManager
} from 'react-native'
import { Colours, FontSizes } from '../GlobalStyles'
import LinearGradient from 'react-native-linear-gradient'

/**
 * A configurable button
 * @example
 * <Button text="Click me" onPress={ () => doSomething()} disabled={false}/>
 */
export default class Button extends Component {

  constructor (props) {
    super(props)
    this.bgColour1 = Colours.buttonColour1
    this.bgColour2 = Colours.buttonColour2

    this.state = {
      bgColour1: this.bgColour1,
      bgColour2: this.bgColour2
    }
    // To prevent double taps
    this.guardDelay = 500
    this.recentlyTouched = false
  }

  componentWillMount () {
    if (this.props.disabled) {
      this.setState({
        bgColour1: this.bgColour1,
        bgColour2: this.bgColour2
      })
    } else {
      this.setState({
        bgColour1: this.bgColour1,
        bgColour2: this.bgColour2
      })
    }
  }

  onPress () {
    if (!this.recentlyTouched) {
      if (this.props.disabled !== true) {
        // Executing immediately here WILL prevent animation
        setTimeout(() => {
          InteractionManager.runAfterInteractions(() => {
            this.props.onPress()
          })
        }, 0)
      } else {
        // If a message was given display it
        if (this.props.disabledText) {
          ToastAndroid.show(this.props.disabledText, ToastAndroid.SHORT)
        }
      }
      // Gaurd
      this.recentlyTouched = true
      // Ungaurd after delay
      setTimeout(() => { this.recentlyTouched = false }, this.guardDelay)
    }
  }

  componentWillReceiveProps (nextProps, nextState) {
    if (this.props.disabled) {
      this.setState({
        bgColour1: Colours.offWhite,
        bgColour2: '#bfbfbf'
      })
    } else {
      this.setState({
        bgColour1: '#fefefe',
        bgColour2: Colours.offWhite
      })
    }
  }

  render () {
    return (
      <View>

        <TouchableHighlight

          underlayColor={Colours.offWhite}

          activeOpacity={this.props.disabled ? 0.4 : 0.2}

          style={[styles.tH, { width: this.props.width || 140, height: this.props.height || 50 }]}

          onPress={() => this.onPress()}
        >

          <View style={styles.view}>
            <LinearGradient colors={[ this.state.bgColour1, this.state.bgColour2 ]} style={styles.lG}>
              <Text style={[styles.text, this.props.style]}>
                {this.props.text}
              </Text>
            </LinearGradient>
          </View>

        </TouchableHighlight>

      </View>
    )
  }
}

const styles = StyleSheet.create({
  tH: {
    marginTop: 4,
    marginBottom: 4,
    borderRadius: 4
  },
  view: {
    flex: 1,
    borderColor: '#a9a9a9',
    borderWidth: 1.5,
    borderStyle: 'solid',
    borderRadius: 4
    // paddingBottom: 0.1
  },
  lG: {
    flex: 1,
    borderRadius: 4,
    alignItems: 'center',
    justifyContent: 'center',
    marginBottom: -0.5
  },
  text: {
    color: Colours.primaryLowlight,
    fontSize: FontSizes.h5,
    textAlign: 'center'
  }
})

Button.propTypes = {
  width: React.PropTypes.number,
  height: React.PropTypes.number,
  text: React.PropTypes.string,
  disabledText: React.PropTypes.string,
  disabled: React.PropTypes.bool,
  onPress: React.PropTypes.func,
  style: React.PropTypes.object
}
