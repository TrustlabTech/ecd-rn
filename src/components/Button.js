/**
 * Early Childhood Development App
 * @copyright 2016 Global Consent Ltd
 * Civvals, 50 Seymour Street, London, England, W1H 7JG
 * @author Alberto Dallaporta <alberto.dallaporta@novalab.io>
 */

'use-strict'

// base libs
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import {
  View,
  Platform,
  StyleSheet,
  TouchableOpacity,
  TouchableNativeFeedback,
} from 'react-native'
import { COLORS } from '../constants' 

export default class Button extends Component {
  render() {

    if (Platform.OS === 'ios' || Platform.Version <= 20 || this.props.nativeFeedback === false)
      return (
        <TouchableOpacity 
          {...this.props}
          style={[styles.iosStyle, this.props.style]}>
            {this.props.children}
        </TouchableOpacity>
      )

    if (Platform.OS === 'android')
      return (
        <TouchableNativeFeedback
          {...this.props}
          style={[styles.androidStyle, this.props.style]}
          background={TouchableNativeFeedback.Ripple(this.props.rippleColor, false)}>
          <View style={[styles.androidStyle, this.props.style]}>
            {this.props.children}
          </View>
        </TouchableNativeFeedback>
    )
  }
}

Button.defaultProps = {
  rippleColor: COLORS.lightGrey,
  nativeFeedback: false,
  style: {},
}

Button.propTypes = {
  ...TouchableOpacity.propTypes,
  ...TouchableNativeFeedback.propTypes,
  rippleColor: PropTypes.string.isRequired,
  nativeFeedback: PropTypes.bool.isRequired,
}

const commonStyle = {
  padding: 6,
}

const styles = StyleSheet.create({
  iosStyle: {
    ...commonStyle,
  },
  androidStyle: {
    ...commonStyle,
  },
})
