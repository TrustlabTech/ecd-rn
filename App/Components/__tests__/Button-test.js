import 'react-native'
import React from 'react'
import Button from '../Button'

import renderer from 'react-test-renderer'

it('renders correctly', () => {
  const tree = renderer.create(
    <Button text="Test" onPress={ () => {}}/>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})