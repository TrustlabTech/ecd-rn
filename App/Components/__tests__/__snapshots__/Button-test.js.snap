exports[`test renders correctly disabled 1`] = `
<View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "borderRadius": 4,
            "marginBottom": 8,
            "marginTop": 8,
          },
          Object {
            "height": 45,
            "width": 140,
          },
        ],
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "borderColor": "#353535",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0.8,
          "flex": 1,
          "paddingBottom": 0.1,
        }
      }>
      <BVLinearGradient
        colors={
          Array [
            4292861919,
            4290756543,
          ]
        }
        end={undefined}
        locations={null}
        start={undefined}
        style={
          Object {
            "alignItems": "center",
            "borderRadius": 4,
            "flex": 1,
            "justifyContent": "center",
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#393939",
                "fontSize": 22,
                "textAlign": "center",
              },
              undefined,
            ]
          }>
          Test
        </Text>
      </BVLinearGradient>
    </View>
  </View>
</View>
`;

exports[`test renders correctly minimal 1`] = `
<View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "borderRadius": 4,
            "marginBottom": 8,
            "marginTop": 8,
          },
          Object {
            "height": 45,
            "width": 140,
          },
        ],
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "borderColor": "#353535",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0.8,
          "flex": 1,
          "paddingBottom": 0.1,
        }
      }>
      <BVLinearGradient
        colors={
          Array [
            4294901502,
            4292861919,
          ]
        }
        end={undefined}
        locations={null}
        start={undefined}
        style={
          Object {
            "alignItems": "center",
            "borderRadius": 4,
            "flex": 1,
            "justifyContent": "center",
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#393939",
                "fontSize": 22,
                "textAlign": "center",
              },
              undefined,
            ]
          }>
          Test
        </Text>
      </BVLinearGradient>
    </View>
  </View>
</View>
`;
